text/0517-io-os-reform.md

```diff
@@ -447,7 +458,173 @@ counts, arguments to `main`, and so on).
+
+### `Read`
+[Read]: #read
+
+The updated `Reader` trait (and its extension) is as follows:
+
+```rust
+trait Read {
+    type Err; // new associated error type
+
+    // unchanged except for error type
+    fn read(&mut self, buf: &mut [u8]) -> Result<uint, Err>;
+}
+
+// extension trait needed for object safety
+trait ReadExt: Read {
+    fn bytes(&mut self) -> Bytes<Self> { ... }
+    fn chars<'r>(&'r mut self) -> Chars<'r, Self, Err> { ... }
+
+    fn read_to_end(&mut self) -> Result<Vec<u8>, Err> { ... }
+    fn read_to_string(&self) -> Result<String, Err> { ... }
+
+    ... // more to come later in the RFC
+}
+impl<R: Read> ReadExt for R {}
+```
+
+Following the
+[trait naming conventions](https://github.com/rust-lang/rfcs/pull/344),
+the trait is renamed to `Read` reflecting the single method it
+provides.
+
+The `read` method should not involve internal looping (even over
+errors like `EINTR`). It is intended to be atomic.
+
+#### Removed methods
+
+The proposed `Read` trait is much slimmer than today's `Reader`. The vast
+majority of removed methods are parsing/deserialization, which were
+discussed above.
+
+The remaining methods (`read_exact`, `read_at_least`, `push`,
+`push_at_least`) were removed for various reasons:
+
+* `read_exact`, `read_at_least`: these are somewhat more obscure
+  conveniences that are not particularly robust due to lack of
+  atomicity.
+
+* `push`, `push_at_least`: these are special-cases for working with
+  `Vec`, which this RFC proposes to replace with a more general
+  mechanism described next.
+
+To provide some of this functionality in a more composition way,
+extend `Vec<T>` with an unsafe method:
+
+```rust
+unsafe fn with_extra(&mut self, n: uint) -> &mut [T];
+```
+
+This method is equivalent to calling `reserve(n)` and then providing a
+slice to the memory starting just after `len()` entries. Using this
+method, clients of `Read` can easily recover the `push` method.
+
+### `Write`
+[Write]: #write
+
+The `Writer` trait is cut down to even smaller size:
+
+```rust
+trait Write {
+    type Err;
+    fn write(&mut self, buf: &[u8]) -> Result<uint, Err>;
+    fn flush(&mut self) -> Result<(), Err> { ... }
+}
+
+trait WriteExt {
+    fn write_all(&mut self, buf: &[u8]) -> Result<(), Err>
+        where Self::Err: WriteAllError { ... };
+    fn write_fmt(&mut self, fmt: &fmt::Arguments) -> Result<(), Err>
+        where Self::Err: WriteAllError { ... };
+}
+
+impl<W: Write> WriteExt for W {}
+
+trait WriteAllError: Sized {
+    fn eof() -> Self;
+    fn interrupted(&self) -> bool;
+}
+```
+
+The biggest change here is to the semantics of `write`. Instead of
+repeatedly writing to the underlying IO object until all of `buf` is
+written, it attempts a *single* write and on success returns the
+number of bytes written. This follows the long tradition of blocking
+IO, and is a more fundamental building block than the looping write we
+currently have.
```

Comment:
The main advantage is that we're not imposing an extra set of semantics -- including handling of intermediate errors -- on top of the underlying `write` call. There are many ways one might want to loop over calls to `write`, and by exposing the underlying `write` we make it possible for anyone to do so (rather than locking down a particular implementation).

On the other hand, by providing `write_all` was also give a convenient "default" loop/retry behavior.